server/src-lib/Hasura/Server/Auth/JWT.hs

```diff
@@ -71,45 +72,62 @@ $(J.deriveJSON J.defaultOptions { J.sumEncoding = J.ObjectWithSingleField
 defaultClaimsFormat :: JWTClaimsFormat
 defaultClaimsFormat = JCFJson
 
-allowedRolesClaim :: T.Text
+allowedRolesClaim :: SessionVariable
 allowedRolesClaim = "x-hasura-allowed-roles"
 
-defaultRoleClaim :: T.Text
+defaultRoleClaim :: SessionVariable
 defaultRoleClaim = "x-hasura-default-role"
 
 defaultClaimsNamespace :: Text
 defaultClaimsNamespace = "https://hasura.io/jwt/claims"
 
-type CustomClaimsMap = Map.HashMap Text J.JSONPath
+data JWTClaimsMapValueG v
+  = JWTClaimsMapJSONPath !J.JSONPath
+  | JWTClaimsMapStatic !v
+  deriving (Show,Eq,Functor,Foldable,Traversable)
+
+instance (J.FromJSON v) => J.FromJSON (JWTClaimsMapValueG v) where
+  parseJSON (J.Object obj) =
+    case Map.lookup "path" obj of
+      Just path ->
+        case path of
+          J.String path' -> either fail (pure . JWTClaimsMapJSONPath) $ parseJSONPath path'
+          _              -> fail "expected a string value for a JSON path"
+      Nothing -> fail "path containing the JSON path not present"
+  parseJSON v = JWTClaimsMapStatic <$> J.parseJSON v
+
+instance (J.ToJSON v) => J.ToJSON (JWTClaimsMapValueG v) where
+  toJSON (JWTClaimsMapJSONPath jsonPath) = J.String . T.pack $ encodeJSONPath jsonPath
+  toJSON (JWTClaimsMapStatic v)          = J.toJSON v
+
+type JWTClaimsMapValue = JWTClaimsMapValueG SessionVariableValue
+type JWTClaimsMapAllowedRoles = JWTClaimsMapValueG [SessionVariableValue]
+
+type CustomClaimsMap = Map.HashMap SessionVariable JWTClaimsMapValue
 
 data JWTClaimsMap
   = JWTClaimsMap
-  { jcmDefaultRole  :: !J.JSONPath
-  , jcmAllowedRoles :: !J.JSONPath
+  { jcmDefaultRole  :: !JWTClaimsMapValue
+  , jcmAllowedRoles :: !JWTClaimsMapAllowedRoles
   , jcmCustomClaims :: !CustomClaimsMap
-  } deriving (Show, Eq)
+  } deriving (Show,Eq)
+
+$(J.deriveToJSON (J.aesonDrop 3 J.snakeCase) ''JWTClaimsMap)
```

Comment:
Don't use `deriveToJSON` here! This will effect the idempotent of  to and from JSON instances. Use the code [here](https://github.com/rakeshkky/graphql-engine/pull/4/files#diff-ea85db56216e10e4ed52755329ec3850L106).